marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/builtins/steps/mapping/entity-services/lib.sjs

```diff
@@ -0,0 +1,155 @@
+      let sourcedFrom = mapProperties[prop].sourcedFrom;
+      if (externalMappingRef || isArray) {
+        let propLine;
+        if (externalMappingRef) {
+          let externalMapping = mappingLib.getMappingWithVersion(externalMappingRef.name, externalMappingRef.version).toObject();
+          let externalEntity = getTargetEntity(externalMapping.targetEntityType);
+          let externalEntityName = getEntityName(externalEntity);
+          propLine = `<${prop}><m:call-template name="${externalEntityName}">
+            <m:with-param name="$context" select="."/>
+          </m:call-template></${prop}>`;
+        } else {
+          propLine = `<${prop}><m:val>. ! ${dataType}(.)</m:val></${prop}>`;
+        }
+        propertyLines.push(`<m:for-each><m:select>$context ! ${sourcedFrom}</m:select>
+            ${propLine}
+          </m:for-each>`);
+      } else {
+        let propLine = `<${prop}><m:val>$context ! ${sourcedFrom} ! ${dataType}(.)</m:val></${prop}>`;
+        if (!isRequired) {
+          propLine = `<m:optional>${propLine}</m:optional>`
+        }
+        propertyLines.push(propLine);
+      }
+    }
+  }
+  return propertyLines.join('\n');
+}
+
+function getRelatedMappings(mapping, related = [mapping]) {
+  // get references to external mappings
+  let externalReferences = mapping.xpath('/properties/*/externalMapping');
+  for (let mappingRef of externalReferences) {
+    let name = mappingRef.name;
+    let version = mapping;
+    if (!related.includes((mapping) => mapping.root.name === name && mapping.root.version === version)) {
+      let externalMapping = mappingLib.getMappingWithVersion(name, version);
+      related.push(externalMapping);
+      related = getRelatedMappings(externalMapping, related);
+    }
+  }
+  return related;
+}
+
+function getTargetEntity(targetEntityType) {
+  if (!entitiesByTargetType[targetEntityType]) {
+    let targetArr = targetEntityType.split('/');
+    let entityName = targetArr[targetArr.length - 1];
+    let tVersion = targetArr[targetArr.length - 2].split('-');
+    let modelVersion = tVersion[tVersion.length - 1];
+    let entityModel = fn.head(mappingLib.getModel(entityName, modelVersion));
+    if (entityModel && (entityModel.constructor.name === "Document" || entityModel.constructor.name === "ObjectNode")) {
+      entityModel = entityModel.toObject();
+    }
+    if(!entityModel) {
+      datahub.debug.log({message: 'Could not find a target entity: ' + targetEntityType, type: 'error'});
+      throw Error('Could not find a target entity: ' + targetEntityType);
+    }
+    entitiesByTargetType[targetEntityType] = entityModel;
+  }
+  return entitiesByTargetType[targetEntityType];
+}
+
+function retrieveFunctionImports() {
+  let customImports = [];
+  let shimURIs = datahub.hubUtils.queryLatest(function() {
+    return Sequence.from([
+      cts.uriMatch('/custom-modules/mapping-functions/*.xsl'),
+      cts.uriMatch('/data-hub/5/mapping-functions/*.xsl')
+    ]);
```

Comment:
The function metadata generated have uri ending in *.xslt and not *.xsl

Reply:
Can we tag these with a collection instead of relying on uri?

It'll make the search much faster, especially if have to do this a lot.

Reply:
They are tagged with collection "http://marklogic.com/entity-services/function-metadata/compiled".

Reply:
Updated to use the collection. Submitted a bug that will need to be fixed in the nightly before custom functions will work, though. https://bugtrack.marklogic.com/52807